shark_turbine/kernel/wave/codegen.py

```diff
@@ -227,22 +248,63 @@ def handle_read(emitter: WaveEmitter, node: fx.Node):
     except ValueError as e:
         raise ValidationError("Malformed arguments") from e
 
-    assert mapping is None, "mapping is not supported yet"
-
     vector_shape = cast_py_literal(emitter, (elements_per_thread,))
     # memory has no IR node yet.
     kb_src, kb_ir_type, kb_py_type = cast_kernel_buffer(emitter, memory)
 
     if not hasattr(node, "index"):
         raise ValidationError("codegen expected read to have index attr.")
 
-    start_indices = []
-    for dim_indexing in node.index:
-        start_indices.append(gen_sympy_index(emitter, node.index[dim_indexing].start))
+    index = node.index
 
     element_type = kb_ir_type.element_type
     vector_type = VectorType.get(vector_shape, element_type)
-    result = vector_d.load(vector_type, kb_src, start_indices)
+    if mapping is None:
+        start_indices = _get_start_indices(emitter, index)
+        result = vector_d.load(vector_type, kb_src, start_indices)
+    else:
+        assert (
+            mapping.is_output_identity()
+        ), "non-dentity output mapping is not supported yet"
+        mem_index = memory.index
+        input_mapping = mapping.map_input_indices(mem_index.keys())
+
+        iters = mapping.iters
+        subs = [(sym, expr.start) for sym, expr in zip(iters.keys(), index.values())]
+
+        input_index = {
+            key: m.subs(subs) for key, m in zip(mem_index.keys(), input_mapping)
+        }
+
+        strides = strides_from_symbolic_shape(IndexingContext.current(), mem_index)
+        offsets = []
+        subs = [(sym, 0) for sym in iters.keys()]
+        for i in range(elements_per_thread):
```

Comment:
Can you explain a bit what is going on here? I was reading through vector.gather's doc, this seems to be the `index_vec` but even there it's not really explained very clearly what the `index_vec` does. 

Also any reason why we are using vector.gather as opposed to say `masked_load`?

Reply:
`vector.gather` is modeled against `llvm.masked.gather` intrinsic https://llvm.org/docs/LangRef.html#llvm-masked-gather-intrinsics

`masked_load` will read a contiguous vector from `start_indices` position, but `gather` will read each individual element using `(start_indices + offsets_vec[i])` formula and construct a vector from them.

`vector.gather` semantics is actually underspecified when you have multidimensional memref as input, as `offsets_vec[i]` elements are always interpreted as 1D but I think using linearized indices should work and it translates straightforwardly to llvm intrinsic.

Reply:
OK that makes sense, so we are using the "offset_vecs[i]" to read/take potentially non-contiguous data that will be contiguous **after** the read. Thanks for the explanation. :)